lib/Sema/CSSimplify.cpp

```diff
@@ -6186,6 +6218,91 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyConformsToConstraint(
   return SolutionKind::Error;
 }
 
+ConstraintSystem::SolutionKind ConstraintSystem::simplifyTransitivelyConformsTo(
+    Type type, Type protocolTy, ConstraintLocatorBuilder locator,
+    TypeMatchOptions flags) {
+  // Since this is a performance optimization, let's ignore it
+  // in diagnostic mode.
+  if (shouldAttemptFixes())
+    return SolutionKind::Solved;
+
+  auto formUnsolved = [&]() {
+    // If we're supposed to generate constraints, do so.
+    if (flags.contains(TMF_GenerateConstraints)) {
+      auto *conformance =
+          Constraint::create(*this, ConstraintKind::TransitivelyConformsTo,
+                             type, protocolTy, getConstraintLocator(locator));
+
+      addUnsolvedConstraint(conformance);
+      return SolutionKind::Solved;
+    }
+
+    return SolutionKind::Unsolved;
+  };
+
+  auto resolvedTy = getFixedTypeRecursive(type, /*wantRValue=*/true);
+  if (resolvedTy->isTypeVariableOrMember())
+    return formUnsolved();
+
+  // If the composition consists of a class + protocol,
+  // we can't check conformance of the argument because
+  // parameter could pick one of the components.
+  if (resolvedTy.findIf(
+          [](Type type) { return type->is<ProtocolCompositionType>(); }))
+    return SolutionKind::Solved;
+
+  auto *protocol = protocolTy->castTo<ProtocolType>()->getDecl();
+
+  // First, let's check whether the type itself conforms,
+  // if it does - we are done.
+  if (TypeChecker::conformsToProtocol(resolvedTy, protocol, DC))
+    return SolutionKind::Solved;
```

Comment:
I think you may want to avoid checking conditional conformance requirements by using `lookupConformance` here and below, as it's possible an implicit conversion changes e.g the element type in an array-to-array or optional-to-optional conversion.

Reply:
I don't think it should for anything besides `String` maybe but yeah, it's might be better to under-check here and let the parameter handle the rest...

Reply:
I mean something like:

```
protocol P {}
extension Optional : P where Wrapped == Any {}

func bar<T : P>(_ x: T) -> T { x }
var i: Int?
let _: Any? = bar(i)
```

Reply:
Yeah, I understand.

Reply:
Just to clarify because the documentation of `lookupConformance` isn't super clear - will this return a valid `ProtocolConformanceRef` when checking if `Int?` conforms to `P` in the above example?

Reply:
Yes, `lookupConformance` would return a conformance for `Optional : P` with conditional requirements list.